nifi-nar-bundles/nifi-cassandra-bundle/nifi-cassandra-processors/src/main/java/org/apache/nifi/processors/cassandra/PutCassandraQL.java

```diff
@@ -0,0 +1,383 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+@SupportsBatching
+@Tags({"cassandra", "cql", "put", "insert", "update", "set"})
+@EventDriven
+@InputRequirement(InputRequirement.Requirement.INPUT_REQUIRED)
+@CapabilityDescription("Execute provided Cassandra Query Language (CQL) statement. The content of an incoming FlowFile "
+        + "is expected to be the CQL command to execute. The CQL command may use the ? to escape parameters. In this "
+        + "case, the parameters to use must exist as FlowFile attributes with the naming convention cql.args.N.type "
+        + "and cql.args.N.value, where N is a positive integer. The cql.args.N.type is expected to be "
+        + "a lowercase string indicating the Cassandra type.")
+@ReadsAttributes({
+        @ReadsAttribute(attribute = "cql.args.N.type",
+                description = "Incoming FlowFiles are expected to be parameterized CQL statements. The type of each "
+                        + "parameter is specified as a lowercase string corresponding to the Cassandra data type (text, "
+                        + "int, boolean, e.g.). In the case of collections, the primitive type(s) of the elements in the "
+                        + "collection should be comma-delimited, follow the collection type, and be enclosed in angle brackets "
+                        + "(< and >), for example set<text> or map<timestamp, int>."),
+        @ReadsAttribute(attribute = "cql.args.N.value",
+                description = "Incoming FlowFiles are expected to be parameterized CQL statements. The value of the "
+                        + "parameters are specified as cql.args.1.value, cql.args.2.value, cql.args.3.value, and so on. The "
+                        + " type of the cql.args.1.value parameter is specified by the cql.args.1.type attribute.")
+})
+public class PutCassandraQL extends AbstractCassandraProcessor {
+
+    static final PropertyDescriptor STATEMENT_TIMEOUT = new PropertyDescriptor.Builder()
+            .name("Max Wait Time")
+            .description("The maximum amount of time allowed for a running CQL select query. Must be of format "
+                    + "<duration> <TimeUnit> where <duration> is a non-negative integer and TimeUnit is a supported "
+                    + "Time Unit, such as: nanos, millis, secs, mins, hrs, days. A value of zero means there is no limit. ")
+            .defaultValue("0 seconds")
+            .required(true)
+            .addValidator(StandardValidators.TIME_PERIOD_VALIDATOR)
+            .build();
+
+    private static final Pattern CQL_TYPE_ATTRIBUTE_PATTERN = Pattern.compile("cql\\.args\\.(\\d+)\\.type");
+
+    // Matches on top-level type (primitive types like text,int) and also for collections (like list<boolean> and map<float,double>)
+    private static final Pattern CQL_TYPE_PATTERN = Pattern.compile("([^<]+)(<([^,>]+)(,([^,>]+))*>)?");
+
+    @Override
+    protected void init(final ProcessorInitializationContext context) {
```

Comment:
While a common pattern in many of our processors, given the fact that its a known issue where _init()_ method is invoked several times, I'd suggest moving _relationships_ and _descriptors_ initialization to a static initializer to avoid unnecessary GC. You can look at AMQP, new JMS as well as SNMP PR.